### PR TITLE
Add registry mirror support for pack build

### PR DESCRIPTION
**Pull request by aemengo**
## Summary
Add support for a registry mirroring  during **pack build** command. The use-case being enterprise environments where public registry access is denied.

If pack's `config.toml` is configured like:
```toml
[registry-mirrors]
  "index.docker.io" = "nexus.mycompany.local/mirror"
```

Then requests from registries`paketobuildpacks/builder` will be turned into `nexus.mycompany.local/mirror/paketobuildpacks/builder`

This PR also adds the config subcommands to manipulate these entries to config.toml.

#### Before
```bash
$ pack build aemengo/hello
full: Pulling from paketobuildpacks/builder
Digest: sha256:2917564c2de52debd1df70d11c014b2708c8974e6e853804958825be84cfb859
Status: Image is up to date for paketobuildpacks/builder:full
...
```
#### After

```bash
$ pack build aemengo/hello
full: Pulling from mirror/paketobuildpacks/builder
Digest: sha256:2917564c2de52debd1df70d11c014b2708c8974e6e853804958825be84cfb859
Status: Image is up to date for nexus.mycompany.local/mirror/dockerhub-mirror/paketobuildpacks/builder:full
...
```

## Caveats
* The user entered `image-tag` is not modified. Not sure if this is desired behavior or not. 🤷🏾‍♂️

## Documentation

- Should this change be documented?
    - [x] Yes, see #___
    - [ ] No

## Related

Resolves https://github.com/buildpacks/pack/issues/821

### CODE REVIEW

File: internal/image/fetcher.go [hunk 1]
```diff
@@ -19,25 +19,49 @@ import (
...
+	return fetcher
 }
 
 var ErrNotFound = errors.New("not found")
 
 func (f *Fetcher) Fetch(ctx context.Context, name string, daemon bool, pullPolicy config.PullPolicy) (imgutil.Image, error) {
+	name, err := pname.TranslateRegistry(name, f.registryMirrors)
+	if err != nil {
+		return nil, err
+	}
```
Comment (aemengo) [inline internal/image/fetcher.go:63]: Just acknowledging that this line isn't tested. 😕

**Comment by jromero** [comment]
The implemented proxy per host or base url is interesting. 

```toml
[registry-mirrors]
  "index.docker.io" = "nexus.mycompany.local/mirror"
  "gcr.io/some/path" = "nexus.mycompany.local/mirror"
```

I'm not sure where that requirement came from (I see it in the example config comment [here](https://github.com/buildpacks/pack/issues/821#issuecomment-785298898)) but mirrors are more commonly applied globally to any and all dependencies. I'd want to validate that this feature request does need a per host/base url type configuration. 


1. If it does, we'll need a wildcard proxy so it works across any and all images. 
    ```toml
    [registry-mirrors]
      "*" = "nexus.mycompany.local/mirror"
    ```

2. If there isn't a current requirement for per host/base url, I would remove it to reduce complexity.


Reference:

- maven < 2.0.5: [classic setup](https://www.jfrog.com/confluence/display/RT12/Configuring+Maven)
- maven >= 2.0.5: [`mirrorOf` wildcard](https://maven.apache.org/guides/mini/guide-mirror-settings.html#advanced-mirror-specification)
- npm: [default registry and scoped packages](https://www.jfrog.com/confluence/display/JFROG/npm+Registry#npmRegistry-SettingtheDefaultRegistry) ([scoped packages](https://docs.npmjs.com/cli/v7/using-npm/scope) are like namespaces - analogous to base url)

**Comment by aemengo** [comment]
@jromero This flexibility for per host/base url came from my experience using [harbor as a proxy cache](https://goharbor.io/docs/2.1.0/administration/configure-proxy-cache/).

At the very least, clients using the proxy cache feature of Harbor cannot set a global mirror.

**Comment by jromero** [comment]
TL;DR: we want both per base urls and wildcard (any).

That makes sense although it does appear to be a harbor specific limitation.

FWIW, the global mirror I was referring to is declared at the client side (pack in this case). It basically says, "for any image, I want you to go look it up at this registry" (ie. Harbor or Artifactory), then it's up to configuration in the registry to look up that image in configured remote (delegated) registries. In Artifactory's case you can setup multiple registries it can delegate/proxy so as a client you want "any" support so that it's up to the Artifactory to do the look up without having to manually map all base urls at the client side.

**Review by dfreilich — CHANGES_REQUESTED** [review]
This is really super for my current use case, where I want to ensure all of our requests are proxied to Artifactory. Really happy to see this work @aemengo; superb as always. 

One very confusing aspect of it for me, in UA, was the fetcher doesn't immediately say that it's pulling from the mirror. I cancelled it a number of times, because I didn't think I had set my mirror appropriately, until I got to the `Status` line and saw that it did work appropriately. If you could make the image mirror explicit from the beginning, that would make the UX for me significantly better. 

```
$ out/pack build test-go -B paketobuildpacks/builder:tiny -p ../paketo-samples/go/mod
tiny: Pulling from paketobuildpacks/builder
b734f5039f3d: Pull complete
...
Digest: sha256:e6e5fcd7c7094ed28f45b98e1528303f0f3304696e6bd737537c838cdd8f463f
Status: Downloaded newer image for <Artifactory-mirror>/paketobuildpacks/builder:tiny
tiny-cnb: Pulling from paketobuildpacks/run
4fb75c639a6b: Pull complete
...
Digest: sha256:b643257f93f44aa4e77c1c746fea14f419e29a2123829c7a228bc7f1b633fe2a
Status: Downloaded newer image for <Artifactory Mirror>/paketobuildpacks/run:tiny-cnb
...
```

**Comment by aemengo** [comment]
@dfreilich

> If we could change the `Pulling from` line to also reflect the registries mirror, that would be a slam-dunk from my end

It looks like we're not writing that line, it's being feed to us by the docker API. I don't know how you still feel about replacing it. To me, it's a bit unsavory.

Maybe a big yellow text `WARNING: using mirror <Artifactory-mirror>` is better?

**Review by dfreilich — APPROVED** [review]
Beautiful work. Thanks!
